### PR TITLE
add support of PLATFORM=raspi4-aarch64

### PR DESCRIPTION
**Pull request by MrRobertYuan**
add support of PLATFORM=raspi4-aarch64
You can run the apps which are independent of virtio-device on Raspberry 4B with features "ARCH=aarch64 PLATFORM=raspi4-aarch64".
The specific command about how to run and debug on raspi4 are written in /docs/raspi4.md

### CODE REVIEW

**Comment by chyyuu** [comment]
Good work!

File: modules/axconfig/src/platform/raspi4-aarch64.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Architecture identifier.
+arch = "aarch64"
+# Platform identifier.
+platform = "raspi4-aarch64"
+
+# Base address of the whole physical memory.
+phys-memory-base = "0x0"
+# Size of the whole physical memory.
+phys-memory-size = "0x800_0000"     # 128M
```
Comment (equation314) [inline modules/axconfig/src/platform/raspi4-aarch64.toml]: 4G ?

File: modules/axhal/src/arch/aarch64/mod.rs [hunk 2]
```diff
@@ -93,3 +93,9 @@ pub fn flush_icache_all() {
 pub fn set_exception_vector_base(vbar_el1: usize) {
     VBAR_EL1.set(vbar_el1 as _);
 }
+
+/// Flushes the data cache at addr
```
Comment (equation314) [inline modules/axhal/src/arch/aarch64/mod.rs]: `Flushes the data cache line (64 bytes) at the given virtual address`

File: modules/axhal/src/arch/aarch64/mod.rs [hunk 3]
```diff
@@ -93,3 +93,9 @@ pub fn flush_icache_all() {
...
+#[inline]
+pub fn flush_dcache_line(addr: usize) {
```
Comment (equation314) [inline modules/axhal/src/arch/aarch64/mod.rs]: `vaddr: VirtAddr`

File: modules/axhal/build.rs [hunk 4]
```diff
@@ -6,7 +6,10 @@ fn main() {
 }
 
 fn gen_linker_script(arch: &str) -> Result<()> {
+    #[cfg(not(feature = "platform-raspi4-aarch64"))]
```
Comment (equation314) [inline modules/axhal/build.rs]: Create different linker scripts for different platform, e.g., `linker_qemu-virt-aarch64.lds`, `linker_raspi4-aarch64.lds`.

File: Makefile [hunk 5]
```diff
@@ -67,7 +67,11 @@ GDB ?= gdb-multiarch
 OUT_DIR ?= $(APP)
 
 APP_NAME := $(shell basename $(APP))
-LD_SCRIPT := $(CURDIR)/modules/axhal/linker_$(ARCH).lds
+ifeq ($(PLATFORM), raspi4-aarch64)
+	LD_SCRIPT := $(CURDIR)/modules/axhal/linker_$(ARCH)_raspi4.lds
```
Comment (equation314) [inline Makefile]: `LD_SCRIPT := $(CURDIR)/modules/axhal/linker_$(PLATFORM).lds`

File: modules/axhal/src/platform/mod.rs [hunk 6]
```diff
@@ -1,5 +1,12 @@
 //! Platform-specific operations.
 
+cfg_if::cfg_if! {
+    if #[cfg(target_arch = "aarch64")]{
+        mod aarch64_common;
+        pub use self::aarch64_common::*;
```
Comment (equation314) [inline modules/axhal/src/platform/mod.rs]: Do not `pub use self::aarch64_common::*`

File: modules/axhal/src/platform/aarch64_common/mod.rs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+pub mod generic_timer;
+pub mod pl011;
+pub mod psci;
+
+#[cfg(feature = "irq")]
+pub mod gic;
+
+#[cfg(feature = "irq")]
```
Comment (equation314) [inline modules/axhal/src/platform/aarch64_common/mod.rs]: Do not define high-level `mod`s here, move these into `qemu_virt_aarch64/mod.rs` and `raspi4_aarch64/mod.rs`.

File: modules/axhal/src/platform/aarch64_common/generic_timer.rs [hunk 8]
```diff
@@ -42,15 +42,15 @@ pub fn set_oneshot_timer(deadline_ns: u64) {
 }
 
 /// Early stage initialization: stores the timer frequency.
-pub(super) fn init_early() {
+pub fn init_early() {
```
Comment (equation314) [inline modules/axhal/src/platform/aarch64_common/generic_timer.rs]: `pub(crate)`

File: modules/axhal/src/platform/aarch64_common/generic_timer.rs [hunk 9]
```diff
@@ -42,15 +42,15 @@ pub fn set_oneshot_timer(deadline_ns: u64) {
...
     let freq = CNTFRQ_EL0.get();
     unsafe {
         CNTPCT_TO_NANOS_RATIO = Ratio::new(crate::time::NANOS_PER_SEC as u32, freq as u32);
         NANOS_TO_CNTPCT_RATIO = CNTPCT_TO_NANOS_RATIO.inverse();
     }
 }
 
-pub(super) fn init_percpu() {
+pub fn init_percpu() {
```
Comment (equation314) [inline modules/axhal/src/platform/aarch64_common/generic_timer.rs]: `pub(crate)`

File: modules/axhal/src/platform/aarch64_common/gic.rs [hunk 10]
```diff
@@ -43,14 +42,14 @@ pub fn dispatch_irq(_unused: usize) {
 }
 
 /// Initializes GICD, GICC on the primary CPU.
-pub(super) fn init_primary() {
+pub fn init_primary() {
```
Comment (equation314) [inline modules/axhal/src/platform/aarch64_common/gic.rs]: `pub(crate)`

File: modules/axhal/src/platform/aarch64_common/pl011.rs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+/// Reads a byte from the console, or returns [`None`] if no input is available.
+pub fn getchar() -> Option<u8> {
+    UART.lock().getchar()
+}
+
+pub fn init() {
+    UART.lock().init();
+    #[cfg(feature = "irq")]
```
Comment (equation314) [inline modules/axhal/src/platform/aarch64_common/pl011.rs]: Split into two functions: `init_early()` and `init()`, and enable IRQs after GIC initialization.

File: modules/axhal/src/platform/raspi4_aarch64/mem.rs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            let start = virt_to_phys((ekernel as usize).into()).align_up_4k();
+            let end = PhysAddr::from(axconfig::PHYS_MEMORY_END).align_down_4k();
+            Some(MemRegion {
+                paddr: start,
+                size: end.as_usize() - start.as_usize(),
+                flags: MemRegionFlags::FREE | MemRegionFlags::READ | MemRegionFlags::WRITE,
+                name: "free memory",
+            })
+        }
+        Ordering::Greater => extern_memory_region_at(idx),
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mem.rs]: `idx - common_memory_regions_num()`

File: modules/axhal/src/platform/raspi4_aarch64/boot.rs
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/boot.rs:1]: `boot.rs` is almost the same as in `qemu_virt_aarch64`, merge into one.

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+pub static CPU_SPIN_TABLE: [PhysAddr; 4] = [
+    PhysAddr::from(0xd8),
+    PhysAddr::from(0xe0),
+    PhysAddr::from(0xe8),
+    PhysAddr::from(0xf0),
+];
+
+/// Starts the given secondary CPU with its boot stack.
+pub fn start_secondary_cpu(cpu_id: usize, stack_top: PhysAddr) {
+    let entry_address = virt_to_phys(VirtAddr::from(modify_stack_and_start as usize)).as_usize();
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: `entry_paddr`

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    unsafe {
+        // set the boot code address of the given secondary CPU
+        let release_ptr = phys_to_virt(CPU_SPIN_TABLE[cpu_id]).as_mut_ptr() as *mut usize;
+        release_ptr.write_volatile(entry_address);
+        crate::arch::flush_dcache_line(release_ptr as usize);
+        // set the boot stack of the given secondary CPU
+        SECONDARY_STACK_TOP = stack_top.as_usize();
+        crate::arch::flush_dcache_line(
+            virt_to_phys(VirtAddr::from(
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: `flush_dcache_line` use virtual address?

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        ldr     x21, [x21]
+        mov     sp, x21
+        b       {start_secondary}",
+        secondary_boot_stack = sym SECONDARY_STACK_TOP,
+        start_secondary = sym _start_secondary,
+        phys_virt_offset = const axconfig::PHYS_VIRT_OFFSET,
+    );
+}
+
+#[no_mangle]
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: No need to use `#[no_mangle]`

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 16]
```diff
@@ -0,0 +1,52 @@
+use crate::mem::{phys_to_virt, virt_to_phys, PhysAddr, VirtAddr};
+use aarch64_cpu::asm;
+
+#[no_mangle]
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: No need to use `#[no_mangle]`

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        // set the boot stack of the given secondary CPU
+        SECONDARY_STACK_TOP = stack_top.as_usize();
+        crate::arch::flush_dcache_line(
+            virt_to_phys(VirtAddr::from(
+                (&SECONDARY_STACK_TOP as *const usize) as usize,
+            ))
+            .as_usize(),
+        );
+    }
+    asm::sev();
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: use full path: `aarch64_cpu::asm::sev()`

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+use aarch64_cpu::asm;
+
+#[no_mangle]
+static mut SECONDARY_STACK_TOP: usize = 0;
+
+extern "C" {
+    fn _start_secondary();
+}
+
+#[no_mangle]
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: Add `#[naked]`, `
#[link_section = ".text.boot"]`. `#[no_mangle]` is no need.

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+unsafe extern "C" fn modify_stack_and_start() {
+    core::arch::asm!("
+        ldr     x21, ={secondary_boot_stack}    // the secondary CPU hasn't set the TTBR1
+        mov     x8, {phys_virt_offset}          // minus the offset to get the phys addr of the boot stack
+        sub     x21, x21, x8
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: `SECONDARY_STACK_TOP` is already use the physical address?

File: modules/axhal/src/platform/raspi4_aarch64/mp.rs [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        ldr     x21, [x21]
+        mov     sp, x21
+        b       {start_secondary}",
```
Comment (equation314) [inline modules/axhal/src/platform/raspi4_aarch64/mp.rs]: `b    _start_secondary` is ok.

**Review by equation314 — CHANGES_REQUESTED** [review]

File: crates/arm_pl011/src/pl011.rs [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    const fn regs(&self) -> &Pl011UartRegs {
+        unsafe { self.base.as_ref() }
+    }
+
+    /// Initializes the Pl011 UART.
+    ///
+    /// It clears all irqs, sets fifo trigger level, enables rx interrupt, enables receives
+    pub fn init(&mut self) {
+        // clear all irqs
+        self.regs().icr.set(0x3ff);
```
Comment (equation314) [inline crates/arm_pl011/src/pl011.rs]: `0x7ff`

File: crates/arm_pl011/src/pl011.rs [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+        if self.regs().fr.get() & (1 << 4) == 0 {
+            Some(self.regs().dr.get() as u8)
+        } else {
+            None
+        }
+    }
+
+    /// IRQ Handler when receiving the data
+    /// It will simply output the char
+    pub fn handle(&mut self) {
```
Comment (equation314) [inline crates/arm_pl011/src/pl011.rs]: Split into two functions: 

1. `ack_interrupts()`: set ICR.
2. `is_receive_interrupt()`: return `pending & (1 << 4) != 0`.

**Pull request closed without merging**